File: foundation/money/money.go
// Package money holds the helpers for displaying amounts. Amounts are always an
// int64 count of nano units plus a currency; storage and arithmetic stay in nano
// units, and only display converts to the major unit.
package money

import (
	"fmt"
	"math/big"
	"regexp"
	"strings"
)

// NanoPerUnit is how many nano units make one major unit. Exported because the
// literal was written out at five call sites, each one a place where a typo in
// the number of zeros would be arithmetically silent.
const NanoPerUnit = 1_000_000_000

// nanoPerUnit keeps the unexported spelling for this file's own arithmetic.
const nanoPerUnit = NanoPerUnit

// FormatNano renders nano units as a major-unit string with two decimals. It
// carries no currency symbol; the caller adds one. The value is truncated rather
// than rounded, and the sign is preserved. The absolute value is taken in uint64
// so that the most negative int64 does not overflow when negated.
func FormatNano(nano int64) string {
	neg := nano < 0
	var abs uint64
	if neg {
		abs = uint64(-(nano + 1)) + 1 // absolute value that is safe for MinInt64
	} else {
		abs = uint64(nano)
	}
	major := abs / nanoPerUnit
	cents := (abs % nanoPerUnit) / (nanoPerUnit / 100)
	sign := ""
	if neg {
		sign = "-"
	}
	return fmt.Sprintf("%s%d.%02d", sign, major, cents)
}

// FormatNanoExact renders nano units at full precision, dropping meaningless
// trailing zeros: 1_500_000_000 is "1.5", 1_000_000_000 is "1", 1 is
// "0.000000001".
//
// This is the wire form. Money crosses the wire as a string, never as a JSON
// number: a number goes through the client's float parser and then no two
// clients agree about the low decimal places.
//
// It is deliberately not FormatNano. That one is the *display* form — two
// decimals, truncated — and the two must not be reachable through one name:
// a caller who wants to show "$1.50" and a caller who has to transmit
// 0.000000001 want different, both-correct answers.
//
// # The sign
//
// Three hand-written copies of this stood in the tree, and two of them were
// wrong for negatives in a way nothing reported, because every call site
// happened to guard its input:
//
//	catalog: -500_000_000 -> "0.-5"   (and -1 -> "0.0000000-1")
//	proxy:   -500_000_000 -> "0.5"    -- the sign silently gone
//
// The second is the worse one: it answers a negative amount with a positive
// number. Both came from taking the fraction of a negative division and
// formatting it as if the sign lived in the whole part — which it does only
// while the absolute value is at least one.
func FormatNanoExact(nano int64) string {
	return FormatScaled(big.NewInt(nano), 9)
}

// FormatScaled renders a scaled integer as a decimal string with no rounding
// and no trailing zeros. `scale` is how many of the digits are fractional, so
// FormatScaled(n, 9) is the nano form.
//
// big.Int rather than int64 because the caller applying several basis-point
// multipliers needs a denominator of 10^(9+4N), and the product overflows an
// int64 long before the result stops being a terminating decimal.
func FormatScaled(n *big.Int, scale int) string {
	if n.Sign() == 0 {
		return "0"
	}
	negative := n.Sign() < 0
	digits := new(big.Int).Abs(n).String()
	if len(digits) <= scale {
		digits = strings.Repeat("0", scale-len(digits)+1) + digits
	}
	cut := len(digits) - scale
	whole, fraction := digits[:cut], strings.TrimRight(digits[cut:], "0")
	out := whole
	if fraction != "" {
		out += "." + fraction
	}
	if negative {
		out = "-" + out
	}
	return out
}

// 十进制字面量的两种接受面。
//
// 此前是三份正则，散在三个包里，且**已经漂了**：`^-?\d+(\.\d+)?$`、
// `^(0|[1-9][0-9]*)(\.[0-9]{1,9})?$`、`^(0|[1-9][0-9]*)(\.[0-9]+)?$`。
// 三者的差别不是笔误，是三个不同的判断——但没有一处说得出自己和另外两处差在哪，
// 而第四个需要校验小数的地方会去抄离它最近的那一份（ADR-0196）。
//
// 收成两个具名函数，各自说明它接受什么：
//
//   - IsSignedDecimal 服务**通用**设置值。它不知道那个数是什么意思，所以既不拒绝
//     负号也不拒绝前导零——一个知道得更少的校验器不该假装知道得更多。
//   - IsPlainDecimal 服务**价格与费率**。这些数进账单，故非负、无前导零，并可选
//     地限制小数位——超出存储精度的位数在别处会被静默舍掉，不如在入口拒绝。
//
// 两者都拒绝科学计数法与两侧空白：一个只通过没人复核过的格式到达网关的费率，
// 正是那种最后差三个数量级的数。
var (
	signedDecimalRE = regexp.MustCompile(`^-?\d+(\.\d+)?$`)
	plainDecimalRE  = regexp.MustCompile(`^(0|[1-9][0-9]*)(\.[0-9]+)?$`)
)

// IsSignedDecimal 判定一个可带符号的十进制字面量。
func IsSignedDecimal(s string) bool { return signedDecimalRE.MatchString(s) }

// IsPlainDecimal 判定一个非负、无前导零的十进制字面量。places > 0 时同时限制
// 小数位数；places <= 0 表示不限。
func IsPlainDecimal(s string, places int) bool {
	if !plainDecimalRE.MatchString(s) {
		return false
	}
	if places <= 0 {
		return true
	}
	dot := strings.IndexByte(s, '.')
	return dot < 0 || len(s)-dot-1 <= places
}

File: web/packages/ui/src/money.ts
/**
 * Money display and input parsing. It lives in the shared layer because both
 * applications need it, and each having its own copy was the root cause this
 * module was created to remove.
 *
 * This is presentation formatting only. Storage and arithmetic are always int64
 * nano units; nothing here takes part in a monetary calculation, it merely turns
 * nano into a human-readable string in main units.
 *
 * Before it existed, a credits page printed raw nano values — a balance shown as
 * 5000000000 — while an input further down the same page took main units and a
 * threshold input next to it took nano: three units in one card. The original
 * formatter was buried in a usage route, so any other page wanting a general
 * money function had to import it from "the usage page", which made no sense.
 */

/** formatNano converts nano to main units using integer arithmetic, keeping
 * floating-point error out of the presentation layer. */
export function formatNano(nano: number): string {
  // Four decimals, then trailing zeros trimmed. It goes through the fixed form
  // so the rounding rule is written once; both used to cut independently, and a
  // rule copied into two places is a rule that eventually differs between them.
  const fixed = formatNanoFixed(nano, 4);
  const [whole, frac = ""] = fixed.split(".");
  const trimmed = frac.replace(/0+$/, "");
  return trimmed ? `${whole}.${trimmed}` : whole!;
}

/**
 * formatNanoFixed is the fixed-decimal variant.
 *
 * `formatNano` strips trailing zeros — 1.5 rather than 1.5000 — which suits
 * running text. Money columns in a table, however, have to line up, and for
 * small values such as unit prices the trailing zeros are what reveal the
 * precision. Four separate local implementations of `(v/1e9).toFixed(n)` had
 * grown up around this need, two using two decimals and two using four.
 */
export function formatNanoFixed(nano: number, digits = 2): string {
  const neg = nano < 0;
  const abs = Math.abs(nano);
  // Round at the display boundary rather than cut. Cutting is not a smaller
  // rounding error, it is a one-directional one: 1234.56789 became 1234.5678,
  // always understating, and always by the same sign. On the pages that show a
  // balance broken into its parts — the whole point of which is that a reader
  // can add the rows up — the cut rows summed to a different number than the cut
  // total, so the page contradicted itself in the one place it was making an
  // arithmetic claim.
  //
  // Integer arithmetic throughout: `scale` is how many nano one displayed unit
  // is worth, so this never divides into a float and back.
  const scale = 10 ** (9 - digits);
  const units = Math.round(abs / scale);
  const perWhole = 10 ** digits;
  const whole = Math.floor(units / perWhole);
  const frac = String(units % perWhole).padStart(digits, "0");
  const s = digits > 0 ? `${whole}.${frac}` : String(whole);
  return neg ? `-${s}` : s;
}

/**
 * formatMoney shows an amount *with its currency*, and is the default in both
 * applications.
 *
 * An amount is an int64 nano value **plus a currency**: the currency is half of
 * what the amount means, and the presentation layer used to drop it. Every call
 * site assembled the pair by hand — one page had even grown its own local money
 * helper — while of 26 formatting calls on the other side only four had a
 * currency anywhere near them. The same number read as "0 USD" in one
 * application and "0.00" in the other, with 33 bare figures measured in total.
 *
 * `digits` means what it does in `formatNanoFixed`: give a digit count for the
 * fixed form, used where columns must align or where trailing zeros carry the
 * precision, and omit it for the trimmed form used in running text.
 *
 * With no currency it prints the number alone rather than assuming one. Guessing
 * a currency is worse than omitting it, because it asserts something to the
 * reader that nothing supports.
 */
export function formatMoney(
  nano: number,
  currency?: string,
  options?: { digits?: number },
): string {
  const amount =
    options?.digits === undefined ? formatNano(nano) : formatNanoFixed(nano, options.digits);
  return currency ? `${amount} ${currency}` : amount;
}

/** nanoToMain converts nano to a main-unit number, for places such as chart axes
 * that need a number rather than a string. */
export function nanoToMain(nano: number): number {
  return nano / 1_000_000_000;
}

/** mainToNano converts main-unit input to nano by padding the string, never
 * going through floating point. */
export function mainToNano(v: string): number {
  const [w = "0", f = ""] = v.trim().split(".");
  return Number(w) * 1_000_000_000 + Number(f.padEnd(9, "0").slice(0, 9));
}

/**
 * formatNanoInput is the lossless main-unit representation: all nine decimals
 * kept, trailing zeros trimmed.
 *
 * Form pre-fill must use it rather than `formatNano`, which keeps only four
 * decimals. Round-tripping through the shorter form means that opening a panel,
 * changing nothing and pressing save silently alters the price:
 *
 *     3333333333 → formatNano "3.3333" → mainToNano 3333300000   ✗ 33333 lost
 *
 * The division of labour: this function for anything editable, `formatNanoFixed`
 * for read-only tables where alignment matters.
 */
export function formatNanoInput(nano: number): string {
  const s = formatNanoFixed(nano, 9);
  return s.includes(".") ? s.replace(/0+$/, "").replace(/\.$/, "") : s;
}

/**
 * normalizeAmountInput tidies a pasted price by stripping the currency symbol
 * and thousands separators.
 *
 * Prices get copied from an upstream pricing page, where they are written as
 * `$3.00` or `1,250`. Without stripping, those are rejected as "not a number" —
 * but this is a difference in formatting, not a mistake by the person typing.
 *
 * The stripped set includes U+FFE5, the fullwidth yen a CJK keyboard actually
 * produces. It is written as an escape rather than literally so that this file
 * carries no fullwidth character of its own.
 */
export function normalizeAmountInput(v: string): string {
  return v
    .trim()
    .replace(/^[$¥\uFFE5]/, "")
    .replace(/,/g, "")
    .trim();
}

/**
 * Validation for money and numeric input.
 *
 * Two real gaps prompted it:
 * - a pricing form used `Number(x) || 0`, so invalid input silently became 0 and
 *   the model in question became free;
 * - a funds form used `Number(x)`, so "abc" became NaN and was serialised to
 *   null on its way to the server.
 *
 * Schemas rather than hand-written conditionals, so that "is this a usable
 * number" is decided once, with one set of messages, shared by both forms.
 */
import { z } from "zod";

const decimal = z
  .string()
  .trim()
  .refine((v) => v !== "" && Number.isFinite(Number(v)), { message: "notANumber" });

/** amountSchema validates a main-unit amount; with `positive` it additionally
 * requires a value greater than zero, as when granting credit. */
export function amountSchema(opts: { positive?: boolean; nonZero?: boolean } = {}) {
  return decimal.superRefine((v, ctx) => {
    const n = Number(v);
    if (opts.positive && n <= 0) ctx.addIssue({ code: "custom", message: "mustBePositive" });
    if (opts.nonZero && n === 0) ctx.addIssue({ code: "custom", message: "mustBeNonZero" });
  });
}

/** intSchema validates a non-negative integer. Prices are stored in nano and
 * markups in basis points, and neither accepts a fraction. */
export const intSchema = decimal.refine((v) => Number.isInteger(Number(v)) && Number(v) >= 0, {
  message: "mustBeNonNegativeInt",
});

/** validate returns a message key for `t()` to resolve, or undefined when the
 * value passes. */
export function validate(schema: z.ZodType<string>, value: string): string | undefined {
  const r = schema.safeParse(value);
  return r.success ? undefined : r.error.issues[0]?.message;
}
